src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+                    SecAlg::RSASHA512 => (MessageDigest::sha512(), 1024 / 8),
+                    _ => unreachable!(),
+                };
+
+                // The key isn't available in either PEM or DER, so use the
+                // direct RSA verifier.
+                let (e, n) = rsa_exponent_modulus(dnskey, min_bytes)?;
+                let e = BigNum::from_slice(&e)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let n = BigNum::from_slice(&n)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let public_key = Rsa::from_public_components(n, e)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let public_key = PKey::from_rsa(public_key)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                Ok(PublicKey::Default(
+                    digest_algorithm,
+                    public_key,
+                    dnskey.flags(),
+                ))
+            }
+            SecAlg::ECDSAP256SHA256 | SecAlg::ECDSAP384SHA384 => {
+                let (digest_algorithm, group_id) = match sec_alg {
+                    SecAlg::ECDSAP256SHA256 => {
+                        (MessageDigest::sha256(), Nid::X9_62_PRIME256V1)
+                    }
+                    SecAlg::ECDSAP384SHA384 => {
+                        (MessageDigest::sha384(), Nid::SECP384R1)
+                    }
+                    _ => unreachable!(),
+                };
+
+                let group = EcGroup::from_curve_name(group_id)
+                    .expect("should not fail");
+                let mut ctx = BigNumContext::new().expect("should not fail");
+
+                // Add 0x4 identifier to the ECDSA pubkey as expected by openssl.
+                let public_key = dnskey.public_key().as_ref();
+                let mut key = Vec::with_capacity(public_key.len() + 1);
+                key.push(0x4);
+                key.extend_from_slice(public_key);
+                let point = EcPoint::from_bytes(&group, &key, &mut ctx)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                let public_key = EcKey::from_public_key(&group, &point)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                public_key
+                    .check_key()
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+
+                Ok(PublicKey::EcDsa(
+                    digest_algorithm,
+                    public_key,
+                    dnskey.flags(),
+                ))
+            }
+            SecAlg::ED25519 => {
+                let public_key = PKey::public_key_from_raw_bytes(
+                    dnskey.public_key().as_ref(),
+                    Id::ED25519,
+                )
+                .map_err(|_| AlgorithmError::InvalidData)?;
+                Ok(PublicKey::NoDigest(public_key, dnskey.flags()))
+            }
+            _ => Err(AlgorithmError::Unsupported),
```

Comment:
The original `src/validate.rs` function `from_dnskey()` could handle ED448 but this function does not. Is that intentional? It seems especially odd given that other functions do handle the ED448 case.